Implementing review comments by @peeech

Br1ght0ne opened the pull request.

Br1ght0ne committed:
Write back to $1 instead of persistence_dir

Also:
- add structopt+paw for CLI

Suggested-by: peeech <peeech@users.noreply.github.com>

Br1ght0ne committed:
Set write perms on copied DNAs and new config

Suggested-by: peeech <peeech@users.noreply.github.com>

Br1ght0ne committed:
Rename update_with -> persist_state_from

Suggested-by: peeech <peeech@users.noreply.github.com>

Br1ght0ne committed:
Retry POST to HAPP2HOST at least once

Suggested-by: peeech <peeech@users.noreply.github.com>

Br1ght0ne committed:
Skip persisting state if old config doesn't exist

Suggested-by: peeech <peeech@users.noreply.github.com>

peeech committed:
update cargoSha256

peeech commented on src/types.rs:
```diff
@@ -97,16 +104,28 @@ impl Configuration {
     /// POST Holo-hosted hApp URLs to resolver
     pub fn update_happ2host(&self) -> Result<()> {
+        const RETRIES: usize = 1;
+        const DELAY: u64 = 1000; // ms
```
DELAY is unused. Did you want to `thread::sleep`?

peeech commented on src/main.rs:
```diff
@@ -38,19 +50,19 @@ fn main() -> Result<()> {
+            )
+        },
+    )?;
+    utils::set_write_permissions(&args.config_path)?;
```
This fails on HPOS with error:
```
Apr 16 15:08:39 hpos-1.holo.host systemd[1]: Starting holochain-conductor.service...
Apr 16 15:08:42 hpos-1.holo.host 36zhjz408pqch353sz110d9hcgpb5b96-unit-script-holochain-conductor-pre-start[3985]: Error: failed to set write permissions on /var/lib/holochain-conductor/conductor-config.toml
Apr 16 15:08:42 hpos-1.holo.host 36zhjz408pqch353sz110d9hcgpb5b96-unit-script-holochain-conductor-pre-start[3985]: Caused by:
Apr 16 15:08:42 hpos-1.holo.host 36zhjz408pqch353sz110d9hcgpb5b96-unit-script-holochain-conductor-pre-start[3985]:     Operation not permitted (os error 1)
```
I am not 100% if `set_write_permissions()` in even necessary. I need to do a bit more research on ownership and permissions in the current version of HPOS. Stay tuned.

peeech replied:
Ok, I got it. Our permission problems stem from the fact that currently `holochain-conductor.service` [dynamically allocates process user](https://github.com/Holo-Host/holo-nixpkgs/blob/develop/modules/services/holochain-conductor.nix#L40). So obviously persistent files will cause permissions issue.

I suggest dealing with user on HPOS level - creating static user `holochain-conductor` and running systemd under this user.

Because we are copying dnas from derivations (which have `r--r--r--`) `set_write_permissions()` is still needed after copying.

Br1ght0ne replied:
Thanks, I didn't think about it. Will address it in PR to `holo-nixpkgs`.

Br1ght0ne replied:
Resolved in https://github.com/Holo-Host/holo-nixpkgs/pull/439/commits/30c323eb48a13fadf8d652d3aa3a38c1ab623477.

peeech reviewed (changes requested):
See comments

peeech commented on src/types.rs:
```diff
@@ -35,6 +35,11 @@ impl Configuration {
     /// Copy all DNA files to `persistence_dir` and update their `file` values.
     pub fn copy_dnas_to_persistence_dir(&mut self) -> Result<()> {
+        let dnas_dir = self.persistence_dir.join("dnas");
+        fs::create_dir(&dnas_dir).with_context(|| {
```
Create dir if not exists
```
if !dnas_dir.is_dir() ...
```

peeech commented on src/types.rs:
```diff
@@ -97,16 +104,28 @@ impl Configuration {
 
     /// POST Holo-hosted hApp URLs to resolver
     pub fn update_happ2host(&self) -> Result<()> {
+        const RETRIES: usize = 1;
+        const DELAY: u64 = 1000; // ms
+        const URL: &str = "https://resolver.holohost.net/update/addHost";
```
there's something fishy about using CONST in function for me. CONST remains in memory for entire duration of program, while you need values declared here only in scope of this function block. wouldn't simple `let` be better? Especially that you do `let mut retries = RETRIES;` below anyway.

Br1ght0ne committed:
Refactor code after review

- Check if `dnas_dir` exists before creating it
- Replace usage of `const` with `let`
- use `delay` in `thread::sleep`
